be/src/vec/functions/array/function_array_sort.h

```diff
@@ -0,0 +1,288 @@
+    template <typename SrcDataType>
+    void _sort_by_permutation(ColumnArray::Offset& prev_offset,
+                              const ColumnArray::Offset& curr_offset,
+                              const SrcDataType* src_data_concrete, const IColumn& src_column,
+                              const NullMapType* src_null_map, IColumn::Permutation& permutation) {
+        for (ColumnArray::Offset j = prev_offset; j < curr_offset - 1; ++j) {
+            if (src_null_map && (*src_null_map)[j]) {
+                continue;
+            }
+            for (ColumnArray::Offset k = j + 1; k < curr_offset; ++k) {
+                if (src_null_map && (*src_null_map)[k]) {
+                    continue;
+                }
+                int result = src_data_concrete->compare_at(permutation[j], permutation[k],
+                                                           src_column, 1);
+                if (result > 0) {
+                    auto temp = permutation[j];
+                    permutation[j] = permutation[k];
+                    permutation[k] = temp;
+                }
+            }
+        }
+        return;
+    }
+
+    template <typename ColumnType>
+    bool _execute_number(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                         IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                         const NullMapType* src_null_map, NullMapType* dest_null_map) {
+        using NestType = typename ColumnType::value_type;
+        const ColumnType* src_data_concrete = reinterpret_cast<const ColumnType*>(&src_column);
+        if (!src_data_concrete) {
+            return false;
+        }
+        const PaddedPODArray<NestType>& src_datas = src_data_concrete->get_data();
+
+        ColumnType& dest_data_concrete = reinterpret_cast<ColumnType&>(dest_column);
+        PaddedPODArray<NestType>& dest_datas = dest_data_concrete.get_data();
+
+        ColumnArray::Offset prev_src_offset = 0;
+        IColumn::Permutation permutation(src_column.size());
+        for (size_t i = 0; i < src_column.size(); ++i) {
+            permutation[i] = i;
+        }
+
+        for (auto curr_src_offset : src_offsets) {
+            // filter and insert null element first
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if (src_null_map && (*src_null_map)[j]) {
+                    DCHECK(dest_null_map != nullptr);
+                    (*dest_null_map).push_back(true);
+                    dest_datas.push_back(NestType());
+                }
+            }
+
+            _sort_by_permutation<ColumnType>(prev_src_offset, curr_src_offset, src_data_concrete,
+                                             src_column, src_null_map, permutation);
+
+            // insert non-null element after sort by permutation
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
```

Comment:
it looks like, the two for-loop can merge to one.
// filter and insert null element first
// insert non-null element after sort by permutation

Reply:
if the array is [1, null, 2, null, 3] and just use one loop, the dest_datas will push_back [1, null, 2, null, 3].

Reply:
got it, array_sort will put all nulls in front of array.